sky/dag.py

```diff
@@ -1,75 +1,253 @@
 """DAGs: user applications to be run."""
-import pprint
 import threading
 import typing
-from typing import List, Optional
+from typing import cast, Dict, List, Optional, Set, Union
+
+import networkx as nx
+
+from sky.utils import common_utils
+from sky.utils import ux_utils
 
 if typing.TYPE_CHECKING:
     from sky import task
 
+TaskOrName = Union['task.Task', str]
+
 
 class Dag:
     """Dag: a user application, represented as a DAG of Tasks.
 
+    This class allows users to define and manage directed acyclic graphs
+    (DAGs) of tasks, representing complex workflows.
+
     Examples:
         >>> import sky
-        >>> with sky.Dag() as dag:
-        >>>     task = sky.Task(...)
+        >>> with sky.Dag(name='my_pipeline') as dag:
+        >>>     task1 = sky.Task(name='task1', ...)
+        >>>     task2 = sky.Task(name='task2', ...)
+        >>>     task1 >> task2
     """
 
-    def __init__(self) -> None:
-        self.tasks: List['task.Task'] = []
-        import networkx as nx  # pylint: disable=import-outside-toplevel
+    def __init__(self, name: Optional[str] = None) -> None:
+        """Initialize a new DAG.
 
+        Args:
+            name: Optional name for the DAG.
+        """
+        self.name = name
+        self._task_name_lookup: Dict[str, 'task.Task'] = {}
+        self.edges: Dict['task.Task', Set['task.Task']] = {}
```

Comment:
I'm a little bit confused why we need to maintain our own edges. doesnt `nx.DiGraph` provide this ?

Reply:
Good catch. Ditched `self.edges` and are now fully using `nx.DiGraph` for edge stuff. Makes sense since we got rid of `self.tasks` too. These help keeping things cleaner.